turbo/snapshotsync/freezeblocks/block_reader.go

```diff
@@ -1250,6 +1253,53 @@ func (r *BlockReader) EventsByBlock(ctx context.Context, tx kv.Tx, hash common.H
 	return result, nil
 }
 
+func (r *BlockReader) EventsById(from uint64, to time.Time, limit int) ([]*heimdall.EventRecordWithTime, error) {
+	view := r.borSn.View()
+	defer view.Close()
+
+	segments := view.Events()
+	var buf []byte
+	var result []*heimdall.EventRecordWithTime
+	stateContract := bor.GenesisContractStateReceiverABI()
+
+	for _, sn := range segments {
+		idxBorTxnHash := sn.Index()
+
+		if idxBorTxnHash == nil {
+			continue
+		}
+		if idxBorTxnHash.KeyCount() == 0 {
+			continue
+		}
+
+		offset := idxBorTxnHash.OrdinalLookup(0)
+		gg := sn.MakeGetter()
+		gg.Reset(offset)
+		for gg.HasNext() {
+			buf, _ = gg.Next(buf[:0])
+
+			raw := rlp.RawValue(common.Copy(buf[length.Hash+length.BlockNum+8:]))
+			event := heimdall.UnpackEventRecordWithTime(stateContract, raw)
```

Comment:
the block reader should not unpack this - it should return []rlp.RlpValue as the other functions.
higher level abstraction should transform from RlpValue to heimdall specific data (you can probably do this in a wrapper func in the simulator)

Reply:
While I agree, the reason I did this was to prevent having to loop over the entire slice again in the calling function. We are already unpacking the event to check for constraints so it would make sense for us to save that instead of having to do the unpack again a short while later.

Reply:
an extra for loop in favour of better abstraction/code organisation is not a big deal unless you are working on ultra low latency
for this particular case I'd say there is bigger fish to fry ie bigger bottlenecks to address before we have to optimise stuff like this